### PR TITLE
Adding historical balance dedup

### PR DESCRIPTION
**Pull request by spiderjako**
Adding a function, which removes every record in the results of historical balance, which is equal to the previous, in this way, there won't be any repeating results

### CODE REVIEW

File: src/functions/blockchainFunctions.js [hunk 1]
```diff
@@ -158,3 +158,26 @@ function topHoldersPercentOfTotalSupply_ (slug, from, to, numberOfHolders) {
...
+  var headers = ['Date', 'Balance']
+
+  var resultsArray = results.map(function (result) {
+    return [
+      formatDatetimeField_(result.datetime),
+      formatNumber_(result.balance)
+    ]
+  })
+
+  for (var i = 1; i < resultsArray.length - 1; i++) {
```
Comment (kalkov) [inline src/functions/blockchainFunctions.js]: Can we use `forEach` here?

**Comment by kalkov** [comment]
@spiderjako Have you deployed and tested the function?

File: src/functions/blockchainFunctions.js [hunk 2]
```diff
@@ -158,3 +158,26 @@ function topHoldersPercentOfTotalSupply_ (slug, from, to, numberOfHolders) {
...
+    if (resultsArray[i][1] === resultsArray[i - 1][1]) {
+      resultsArray.splice(i, 1)
+      i--
```
Comment (IvanIvanoff) [inline src/functions/blockchainFunctions.js]: **never** mutate the loop variable inside its body.
This problem can much better be handled by building a new array. Something like this:
```js
let deduped = []
arr.forEach(function(el) {  if deduped[deduped.length - 1] != el { unique.push(el) }})
```

File: src/functions/blockchainFunctions.js (hunk 1 above)
Comment (spiderjako) [reply src/functions/blockchainFunctions.js]: Yes, I've reworked it now

File: src/functions/blockchainFunctions.js (hunk 2 above)
Comment (spiderjako) [reply src/functions/blockchainFunctions.js]: You're right, it does look a bit fuzzy this way, it is with a forEach cycle now.

File: src/functions/blockchainFunctions.js [hunk 3]
```diff
@@ -158,3 +158,26 @@ function topHoldersPercentOfTotalSupply_ (slug, from, to, numberOfHolders) {
...
+  var resultsArray = results.map(function (result) {
+    return [
+      formatDatetimeField_(result.datetime),
+      formatNumber_(result.balance)
+    ]
+  })
+
+  var deduped = []
+  resultsArray.forEach(function (el, index) {
+    if (deduped.length === 0 || resultsArray.length - 1 === index || deduped[deduped.length - 1][1] !== el[1]) {
```
Comment (IvanIvanoff) [inline src/functions/blockchainFunctions.js]: Why is all that needed? The code I shared works:
```js
let deduped = []
arr.forEach(function(el){  if(deduped[deduped.length - 1] != el) deduped.push(el) })
```

Comment (IvanIvanoff) [reply src/functions/blockchainFunctions.js]: Hm.. or it's needed so the `...[1]` does not raise

**Comment by kalkov** [comment]
Regenerate the docs

File: test/integration/sanHistoricalBalanceDedup.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  })
+
+  it('has positive balances', () => {
+    const results = san.SAN_HISTORICAL_BALANCE_DEDUP(slug, from, to, ethAddress)
+    for (let i = 1; i < results.length; i++) {
+      expect(results[i][1]).to.be.within(0, Infinity)
+    }
+  })
+
+  it('returns at least two records', () => {
```
Comment (IvanIvanoff) [inline test/integration/sanHistoricalBalanceDedup.js]: add comment why `at least two` is required

File: src/functions/blockchainFunctions.js (hunk 3 above)
Comment (valo) [reply src/functions/blockchainFunctions.js]: Why there are duplicates in the result?

Comment (IvanIvanoff) [reply src/functions/blockchainFunctions.js]: The data is returned for every `interval`. If the balance does not change it has the same value as the previous interval. The feature requested (I think by @tzanko-matev) was to be able to fetch only the balances when they change

Comment (valo) [reply src/functions/blockchainFunctions.js]: Isn't it possible to make the deduplication with a function in the Sheets?

Comment (spiderjako) [reply src/functions/blockchainFunctions.js]: After a little digging, there is actually a function =UNIQUE(range) which deduplicates a row

Comment (spiderjako) [reply src/functions/blockchainFunctions.js]: But it won't deduplicate the datetimes next to the results, at least, that won't be so easy. It should be easy for the users to get the deduplicated data IMO and this function does it well.

Comment (IvanIvanoff) [reply src/functions/blockchainFunctions.js]: Note that we do not seek unique but removing only the equal neighbors like this [1,2,2,1,1,2] => [1,2,1,2], I am not sure if such spreadsheet function is available out of the box

**Pull request merged**